### PR TITLE
[WIP] worst case analysis of stack usage

### PR DESCRIPTION
**Pull request by jlecoeur**
This is an experiment on worst case analysis of stack usage. 
The idea would be to compute at compile time the maximum stack usage of apps. The resulting number can be compared to the allocated stack and checked by CI.
 
I added the gcc compilation flag `-fstack-usage` which creates one file with extension `.su` per object file. The `.su` files contain the stack usage per function. I used a perl script `avstack.pl` (originally developped for avr) to walk the call tree and compute the worst case stack usage for each function.
It does not fully work yet. I added sample output of the script for target `px4fmu-v2_default`.

TODO:
- [ ] Check the error messages in file `wca_stack_usage_errors.txt` (see diff)
- [ ] Compare values of stack usage returned in `wca_stack_usage.txt` to what is returned by `top`

If this is fixable, then there will be additional items:
- [ ] add compiler flags only when using gcc
- [ ] update CI to retrieve allocated stack size for every application and compare it with the output of the script

@dagar have you tried something like that already?

### CODE REVIEW

**Comment by dagar** [comment]
Awesome! Yes, I've played with something similar. I recall hitting some difficulties with the random perl script I found online and had to move on.

It would be great if we could roll this into something called from the Makefile. Something like `make stack_check` that runs the script on `px4fmu-v3_default` and compares to the value grepped per module. Then have it run in CI potentially causing the build to fail once we have some confidence it can be trusted.

@davids5 FYI

File: cmake/nuttx/px4_impl_nuttx.cmake [hunk 1]
```diff
@@ -573,10 +573,12 @@ function(px4_os_add_flags)
 	set(added_c_flags
 		-nodefaultlibs
 		-nostdlib
+		-fstack-usage
```
Comment (dagar) [inline cmake/nuttx/px4_impl_nuttx.cmake]: -fcallgraph-info?

**Comment by davids5** [comment]
@jlecoeur @dagar - I would love to see this work. Even if it can not give us an absolute correct value to use as stack sizes. Just a "This PR added/subtracted N" would be a great metric to have and use to guide a stack size adjustment. That value can then be used and tested with HW stack checking enabled (if we can get full code coverage) on a test rig.

**Comment by dagar** [comment]
Good point, a delta would be fantastic.

**Comment by jlecoeur** [comment]
> I recall hitting some difficulties with the random perl script I found online

That was probably the same script! Here is where I found it: http://www.dlbeer.co.nz/oss/avstack.html

As written in the article, most of the errors might be because the output of -stack-usage does not use mangled names. I will try to compile gcc with [this patch](http://www.dlbeer.co.nz/downloads/gcc-4.9.2-mangled-stack-usage.patch).

I added a makefile target to run the script on px4fmu-v3_default.

**Comment by jlecoeur** [comment]
Here is the output of `make check_task`. 
The first script uses the *.su files generated by gcc for each object file. The second script uses only the elf file. 
```
$ make check_stack
ninja: no work to do.
Checking worst case stack usage with avstack.pl ...

  Func                               Cost    Frame   Height
------------------------------------------------------------------------
> sdlog2_thread_main                 1888     1068        9
> hardfault_log_main                  932      460        4
> dumpfile_main                       844      636        3
  sdlog2_start_log                    820      220        8
  sdlog2_stop_log                     796      196        8
> logwriter_thread                    756      164        8
> top_main                            728      228        4
> mtd_main                            672       60        6

Checking worst case stack usage with checkstack.pl ...

Top 10:
local_position_estimator_thread_main(int, char**) [firmware_nuttx]:5600
BlockLocalPositionEstimator::predict() [firmware_nuttx]:3672
position_estimator_inav_thread_main(int, char**) [firmware_nuttx]:3284
Ekf::predictCovariance() [firmware_nuttx]:              3268
Ekf::fuseOptFlow() [firmware_nuttx]:            3140
AttPosEKF::CovariancePrediction(float) [firmware_nuttx]:        2960
Ekf::fuseMag() [firmware_nuttx]:                        2836
Ekf::fuseSideslip() [firmware_nuttx]:           2836
Ekf::fuseDrag() [firmware_nuttx]:                       2644
Ekf::fuseHeading() [firmware_nuttx]:            2484

Symbols with 'main', 'thread' or 'task':
local_position_estimator_thread_main(int, char**) [firmware_nuttx]:5600
position_estimator_inav_thread_main(int, char**) [firmware_nuttx]:3284
commander_thread_main(int, char**) [firmware_nuttx]:            1612
segway_thread_main(int, char**) [firmware_nuttx]:               1292
sdlog2_thread_main [firmware_nuttx]:                    996
fixedwing_control_thread_main(int, char**) [firmware_nuttx]:    860
uORBTest::UnitTest::pubsublatency_main() [firmware_nuttx]:620
dumpfile_main [firmware_nuttx]:                         612
rover_steering_control_thread_main(int, char**) [firmware_nuttx]:580
Mavlink::task_main(int, char**) [firmware_nuttx]:               540
MavlinkReceiver::receive_thread(void*) [firmware_nuttx]:540
MK::task_main() [firmware_nuttx]:                       536
listener_main [firmware_nuttx]:                         492
snapdragon_rc_pwm::task_main(int, char**) [firmware_nuttx]:     428
hardfault_log_main [firmware_nuttx]:                    424
AttitudeEstimatorQ::task_main() [firmware_nuttx]:       372
pwm_main [firmware_nuttx]:                              316
FixedwingAttitudeControl::task_main() [firmware_nuttx]:268
BottleDrop::task_main() [firmware_nuttx]:               268
CameraFeedback::task_main() [firmware_nuttx]:   264
Navigator::task_main() [firmware_nuttx]:                228
top_main [firmware_nuttx]:                              192
TemperatureCalibration::task_main() [firmware_nuttx]:188
frsky_telemetry_thread_main [firmware_nuttx]:           172
oreoled_main [firmware_nuttx]:                          164
px4_simple_app_main [firmware_nuttx]:                   152
PX4IO::task_main() [firmware_nuttx]:                    148
GroundRoverPositionControl::task_main() [firmware_nuttx]:144
vmount_thread_main(int, char**) [firmware_nuttx]:               140
FixedwingPositionControl::task_main() [firmware_nuttx]:140
StateMachineHelperTest::mainStateTransitionTest() [firmware_nuttx]:124
hott_telemetry_thread_main(int, char**) [firmware_nuttx]:       116
logwriter_thread [firmware_nuttx]:                      116
PWMSim::task_main() [firmware_nuttx]:           108
px4_mavlink_debug_main [firmware_nuttx]:                108
uORBTest::UnitTest::pub_test_multi2_main() [firmware_nuttx]:104
esc_calib_main [firmware_nuttx]:                        100
sd_bench_main [firmware_nuttx]:                         92
hott_sensors_thread_main(int, char**) [firmware_nuttx]: 84
uORBTest::UnitTest::pub_test_queue_main() [firmware_nuttx]:84
commander_main [firmware_nuttx]:                        84
MulticopterAttitudeControl::task_main() [firmware_nuttx]:84
GroundRoverAttitudeControl::task_main() [firmware_nuttx]:76
nshterm_main [firmware_nuttx]:                          72
adc_main [firmware_nuttx]:                              68
VtolAttitudeControl::task_main() [firmware_nuttx]:      68
ex_hwtest_main [firmware_nuttx]:                        68
AttitudePositionEstimatorEKF::task_main() [firmware_nuttx]:68
sdlog2_main [firmware_nuttx]:                           64
px4io_main [firmware_nuttx]:                            60
task_main [firmware_nuttx]:                             60
ver_main [firmware_nuttx]:                              56
config_main [firmware_nuttx]:                           52
pwm_out_sim_main [firmware_nuttx]:                      48
led_control_main [firmware_nuttx]:                      48
motor_ramp_thread_main(int, char**) [firmware_nuttx]:   44
bl_update_main [firmware_nuttx]:                        40
mtd_main [firmware_nuttx]:                              40
mkblctrl_main [firmware_nuttx]:                         36
MulticopterPositionControl::task_main() [firmware_nuttx]:36
pwm_input_main [firmware_nuttx]:                        28
gpio_led_main [firmware_nuttx]:                         28
hott_telemetry_main [firmware_nuttx]:                   20
IridiumSBD::main_loop(int, char**) [firmware_nuttx]:    20
ms5611_main [firmware_nuttx]:                           20
PX4FMU::task_spawn(int, char**) [firmware_nuttx]:               20
tap_esc_main [firmware_nuttx]:                          20
motor_test_main [firmware_nuttx]:                       20
reboot_main [firmware_nuttx]:                           20
uavcan_main [firmware_nuttx]:                           20
task_main(int, char**) [firmware_nuttx]:                        20
dataman_main [firmware_nuttx]:                          20
px4_task_spawn_cmd [firmware_nuttx]:                    20
up_taskdump [firmware_nuttx]:                           20
local_position_estimator_thread_main(int, char**) [firmware_nuttx]:4
```
For `sdlog2_thread_main`, there is a ratio 2 between the estimates provided by the two methods, and ratio 6.5 for `logwriter_main`...

**Comment by TSC21** [comment]
Jenkins test this please.

**Comment by TSC21** [comment]
@jlecoeur maybe we can try to fix the indentation of the output?

**Comment by jlecoeur** [comment]
@TSC21 the output will not stay as is. I want to show only the tasks, and display allocated stack vs used stack.
For this I still have to grep the calls to `px4_task_pawn_cmd` and match it with the symbols above.

**Comment by dagar** [comment]
Interesting, we should take a closer look at the corresponding .su files.

Pretty minor, but could you move the tools to a single sub directory in `Tools/`? It's getting pretty crowded/confusing in there. Otherwise we should merge this and continue experimenting.

**Comment by jlecoeur** [comment]
@dagar done!

**Pull request merged**